client/connection_pool.go

```diff
@@ -0,0 +1,269 @@
+		pool:            make([]conn, 0, opts.GetMaxConnectionCount()),
+		poolLen:         0,
+		connectRand:     rand.NewSource(seed),
+		healthCheckRand: rand.NewSource(seed + 1),
+	}
+
+	go p.connectEvery(connectEvery, connectStutter)
+
+	go p.healthCheckEvery(healthCheckEvery, healthCheckStutter)
+
+	return p
+}
+
+func (p *connPool) GetConnectionCount() int {
+	p.RLock()
+	poolLen := p.poolLen
+	p.RUnlock()
+	return int(poolLen)
+}
+
+func (p *connPool) NextClient() (rpc.TChanNode, error) {
+	p.RLock()
+	if p.closed {
+		p.RUnlock()
+		return nil, errConnectionPoolClosed
+	}
+	if p.poolLen < 1 {
+		p.RUnlock()
+		return nil, errConnectionPoolHasNoConnections
+	}
+	n := atomic.AddInt64(&p.used, 1)
+	conn := p.pool[n%p.poolLen]
+	p.RUnlock()
+	return conn.client, nil
+}
+
+func (p *connPool) Close() {
+	p.Lock()
+	if p.closed {
+		p.Unlock()
+		return
+	}
+	p.closed = true
+	p.Unlock()
+
+	for i := range p.pool {
+		p.pool[i].channel.Close()
+	}
+}
+
+func (p *connPool) connectEvery(interval time.Duration, stutter time.Duration) {
+	log := p.opts.GetLogger()
+
+	for {
+		p.RLock()
+		closed := p.closed
+		poolLen := int(p.poolLen)
+		p.RUnlock()
+		if closed {
+			return
+		}
+
+		target := p.opts.GetMaxConnectionCount()
+		if poolLen >= target {
+			// No need to spawn connections
+			time.Sleep(interval + randStutter(p.connectRand, stutter))
+			continue
+		}
+
+		var wg sync.WaitGroup
+		address := p.host.Address()
+		for i := 0; i < target-poolLen; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				channel, err := tchannel.NewChannel(channelName, p.opts.GetChannelOptions())
+				if err != nil {
+					log.Warnf("construct channel error to %s: %v", address, err)
+					return
+				}
+
+				endpoint := &thrift.ClientOptions{HostPort: address}
+				thriftClient := thrift.NewClient(channel, node.ChannelName, endpoint)
+				client := rpc.NewTChanNodeClient(thriftClient)
+
+				// Force a request to open the connection
+				ctx, _ := thrift.NewContext(p.opts.GetHostConnectTimeout())
+				result, err := client.Health(ctx)
+				if err != nil {
+					log.Warnf("could not connect to %s: %v", address, err)
+					return
+				}
+				if !result.Ok {
+					log.Warnf("could not connect to unhealthy host %s: %s", address, result.Status)
+					return
+				}
+
+				p.Lock()
+				if !p.closed {
+					p.pool = append(p.pool, conn{channel, client})
+					p.poolLen = int64(len(p.pool))
+				}
+				p.Unlock()
+			}()
+		}
+
+		wg.Wait()
+
+		time.Sleep(interval + randStutter(p.connectRand, stutter))
+	}
+}
+
+func (p *connPool) healthCheckEvery(interval time.Duration, stutter time.Duration) {
+	log := p.opts.GetLogger()
+	nowFn := p.opts.GetNowFn()
+
+	for {
+		p.RLock()
+		closed := p.closed
+		poolLen := int(p.poolLen)
+		p.RUnlock()
+		if closed {
+			return
+		}
+
+		start := nowFn()
+		deadline := start.Add(interval + randStutter(p.healthCheckRand, stutter))
+		// As this is the only loop that removes connections it is safe to loop upwards
```

Comment:
Health checking takes a long time (due to network) hence do not want to RLock.  Having said that I could simple just use a for loop of `for i := 0; i < poolLen; i++` to avoid the range clause potentially invalidating.